### ISSUE TITLE
Android: dlopen failed: cannot locate symbol "stderr" referenced by "libgmp.so"

### ISSUE THREAD

**Issue by rubenwardy** [minetest/minetest#7963]
Issue reported by @VanessaE. I also recall @MoNTE48 talking about this

```
V/SmartFaceService - 3rd party pause(  687): onReceive [android.intent.action.ACTIVITY_STATE/net.minetest.minetest/create]
D/dalvikvm(16214): Trying to load lib /data/app-lib/net.minetest.minetest-1/libopenal.so 0x428dca70
D/dalvikvm(16214): Added shared lib /data/app-lib/net.minetest.minetest-1/libopenal.so 0x428dca70
D/dalvikvm(16214): Trying to load lib /data/app-lib/net.minetest.minetest-1/libogg.so 0x428dca70
D/dalvikvm(16214): Added shared lib /data/app-lib/net.minetest.minetest-1/libogg.so 0x428dca70
D/dalvikvm(16214): No JNI_OnLoad found in /data/app-lib/net.minetest.minetest-1/libogg.so 0x428dca70, skipping init
D/dalvikvm(16214): Trying to load lib /data/app-lib/net.minetest.minetest-1/libvorbis.so 0x428dca70
D/dalvikvm(16214): Added shared lib /data/app-lib/net.minetest.minetest-1/libvorbis.so 0x428dca70
D/dalvikvm(16214): No JNI_OnLoad found in /data/app-lib/net.minetest.minetest-1/libvorbis.so 0x428dca70, skipping init
D/dalvikvm(16214): No JNI_OnLoad found in /system/lib/libssl.so 0x428dca70, skipping init
D/dalvikvm(16214): No JNI_OnLoad found in /system/lib/libcrypto.so 0x428dca70, skipping init
D/dalvikvm(16214): Trying to load lib /data/app-lib/net.minetest.minetest-1/libgmp.so 0x428dca70
E/dalvikvm(16214): dlopen("/data/app-lib/net.minetest.minetest-1/libgmp.so") failed: dlopen failed: cannot locate symbol "stderr" referenced by "libgmp.so"...
W/dalvikvm(16214): Exception Ljava/lang/UnsatisfiedLinkError; thrown while initializing Lnet/minetest/minetest/MtNativeActivity;
W/dalvikvm(16214): Class init failed in newInstance call (Lnet/minetest/minetest/MtNativeActivity;)
D/AndroidRuntime(16214): Shutting down VM
W/dalvikvm(16214): threadid=1: thread exiting with uncaught exception (group=0x41793898)
E/AndroidRuntime(16214): FATAL EXCEPTION: main
E/AndroidRuntime(16214): java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "stderr" referenced by "libgmp.so"...
E/AndroidRuntime(16214): 	at java.lang.Runtime.loadLibrary(Runtime.java:362)
E/AndroidRuntime(16214): 	at java.lang.System.loadLibrary(System.java:525)
E/AndroidRuntime(16214): 	at net.minetest.minetest.MtNativeActivity.<clinit>(MtNativeActivity.java:88)
E/AndroidRuntime(16214): 	at java.lang.Class.newInstanceImpl(Native Method)
E/AndroidRuntime(16214): 	at java.lang.Class.newInstance(Class.java:1130)
E/AndroidRuntime(16214): 	at android.app.Instrumentation.newActivity(Instrumentation.java:1078)
E/AndroidRuntime(16214): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2210)
E/AndroidRuntime(16214): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2349)
E/AndroidRuntime(16214): 	at android.app.ActivityThread.access$700(ActivityThread.java:159)
E/AndroidRuntime(16214): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1316)
E/AndroidRuntime(16214): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(16214): 	at android.os.Looper.loop(Looper.java:137)
E/AndroidRuntime(16214): 	at android.app.ActivityThread.main(ActivityThread.java:5419)
E/AndroidRuntime(16214): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime(16214): 	at java.lang.reflect.Method.invoke(Method.java:525)
E/AndroidRuntime(16214): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1187)
E/AndroidRuntime(16214): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
E/AndroidRuntime(16214): 	at dalvik.system.NativeStart.main(Native Method)
E/MP-Decision( 1812): num online cores: 2 reqd : 3 available : 4 rq_depth:3.200000 hotplug_avg_load_dw: 114
E/MP-Decision( 1812): UP cpu:1 core_idx:1 Nw:1.900000 Tw:140 total_time_up:0.000000
E/MP-Decision( 1812): UP cpu:2 core_idx:2 Nw:2.700000 Tw:90 total_time_up:98.000000
E/MP-Decision( 1812): num online cores: 3 reqd : 2 available : 4 rq_depth:1.000000 hotplug_avg_load_dw: 102
E/MP-Decision( 1812): DOWN cpu:2 core_idx:2 Ns:2.100000 Ts:240 total_time_down:243.000000
D/dalvikvm(  687): GC_FOR_ALLOC freed 2638K, 17% free 77824K/93368K, paused 151ms, total 152ms
I/WindowManager(  687): Screenshot Window{42d02df8 u0 Starting net.minetest.minetest} was all black! mSurfaceLayer=21040 minLayer=21040 maxLayer=21040
D/SSRMv2:CustomFrequencyManagerService(  687): releaseDVFSLockLocked : Getting Lock type frm List : DVFS_MIN_LIMIT  frequency : 1350000  uid : 1000  pid : 687  tag : ACTIVITY_RESUME_BOOSTER@9
I/power   (  687): *** release_dvfs_lock : lockType : 1 
D/CrashAnrDetector(  687): processName: net.minetest.minetest
D/CrashAnrDetector(  687): broadcastEvent : net.minetest.minetest data_app_crash
W/LicenseLogService(  687): log() is called by non admin
```

**Comment by paramat** [comment]
Blocker?

**Comment by sfan5** [comment]
stderr is only available for api level 23+
`sysroot/usr/include/stdio.h:extern FILE* stderr __INTRODUCED_IN(23);`

we target the much lower 14
`API = 14` ([build/android/Makefile](https://github.com/minetest/minetest/blob/master/build/android/Makefile#L9))

**Comment by rubenwardy** [comment]
I'm unsure of the affect/reach of this, which is why I didn't mark as blocker begin with

**Comment by IhrFussel** [comment]
nerzhul likely compiled this bugfix release with a lib that uses 'stderr' (that's not supported on Android <6 as already mentioned above).

Cause it DOES work on my phone (Android 7.0) and doesn't on old Android versions.

**Comment by stujones11** [comment]
> I'm unsure of the affect/reach of this

It seems limited to Android 4 and somehow related to switching to a shared STL (libc++) as this was not a problem in 0.4.x build with static STL and gnustl.

One possible solution here is to drop the crypto support on android (as MultiCraft does) however, @nerzhul disapproves of this option. Otherwise we need to figure out our true minimum target. I am pretty sure it is lower than api 23 but I may be wrong, I only have 7 & 8 available to test with.

**Comment by IhrFussel** [comment]
@stujones11 someone forgot to mention that this ONLY affects the bugfix apk @nerzhul made in Nov.

I asked @VanessaE again.

**Comment by stujones11** [comment]
@IhrFussel Sorry, my comment was not clear, the STL changes were backported to 0.4.17.x, I have updated my comment to be more clear.

**Comment by IhrFussel** [comment]
@stujones11 so that should mean EVERY apk since that commit breaks on Android <6 ?

**Comment by IhrFussel** [comment]
@stujones11 that is weird.

@VanessaE says that your 5.0.0-dev from the forum works while 0.4.X doesn't.

**Comment by stujones11** [comment]
> says that your 5.0.0-dev from the forum works while 0.4.X doesn't.

That, surely can't be possible, wtf

**Comment by IhrFussel** [comment]
@stujones11 IRC log:

<IhrFussel> VanessaE, just to be sure: You tested the stable 0.4.17.1 from the play store, 0.4.17.2 linked somewhere and 0.4.17.1-3 (the most recent posted by nrz) and none of them startup?
<VanessaE> ihrfussel: 0.4.17.2 in the play store, 0.4.17.1 on minetest.net, and that 0.4.17.1-3 apk
<IhrFussel> And the 5.0.0-dev from stu works correctly... how can THAT be
<VanessaE> yes.

**Comment by stujones11** [comment]
Very strange. I this nerz plans to upload an official pre-release later for testing. If that works then this issue is not really valid, except we do still need to determine our true minimum working android version.

**Comment by IhrFussel** [comment]
@stujones11 Supposedly ALL 3 0.4.X builds @VanessaE checked were compiled by @nerzhul 

So it looks like he uses unsupported libs when compiling?

**Comment by stujones11** [comment]
> So it looks like he uses unsupported libs when compiling?

No, although the lastest one has libs built in debug mode it is unlikely to be because of that.

**Comment by IhrFussel** [comment]
But it would explain why your build works.

**Comment by stujones11** [comment]
> So it looks like he uses unsupported libs when compiling?

By latest one I meant the playstore release, the latest GH release was supposedly re-built, though I haven't checked it yet personally.

**Comment by IhrFussel** [comment]
Yes, but I mean nrz might compile the release apks (and also for testing) in a wrong environment which requires Android 6+.

Cause your builds (I guess different setup) work for her.

You need to explain to me why that couldn't be the case.

**Comment by stujones11** [comment]
> You need to explain to me why that couldn't be the case.

We are both using the same toolchain and compiler, the libs are all built from source.

**Comment by stujones11** [comment]
> -D__ANDROID_API__=14

This is already done by the Makefile.

**Comment by IhrFussel** [comment]
You CANNOT drop Android 4 and Android 5 ... too many users still have devices with those.

There must be a way to support Android 4+.

**Comment by VanessaE** [comment]
Why would the choics of compiler matter for publishing on the play store?

**Comment by stujones11** [comment]
> Why would the choics of compiler matter for publishing on the play store?

Because the are no longer supported in the latest ndks so you will need to use to meet the requirements. Also, gcc did not compile sine the c++11 changes. It was fixed by switching to clang.

**Comment by VanessaE** [comment]
Yes, you 5.0.0 build runs, though I can't use it since my servers all run 0.4.17.1.

**Comment by stujones11** [comment]
Great, then then it is not the stl issue in your case. Which version of android are you using?

**Comment by VanessaE** [comment]
Android 4.3 on a Galaxy S4

**Comment by paramat** [comment]
Note that apparently 0.4.17.1-3 in the play store still has the 'invalid enum' bug, so is still built incorrectly with Irrlicht in debug mode.

**Comment by MoNTE48** [comment]
I spent a lot of time, but I could not build minetest with c++11 for it to work on Android 4. This is a stl problem. But I could not find a solution.

**Comment by nerzhul** [comment]
@paramat i dropped the whole build dir and redo the release process, it's anoying

**Comment by stujones11** [comment]
> I could not build minetest with c++11 for it to work on Android 4

I was also beginning to think this was the case, however, my dev build is working for VanessaE using Android 4.3 (api 18) Perhaps it only effects earlier 4.x versions?

Interestingly, I can't seem to get all the deps to build with the lastest ndk-18 using anything lower than api 16 (androd 4.1) The problems there were also stl related, iirc.

**Comment by paramat** [comment]
> Note that apparently 0.4.17.1-3 still has the 'invalid enum' bug, so is still built incorrectly with Irrlicht in debug mode.

To be clear, i mentioned that in case that was the cause of this bug, which it probably isn't.
nerzhul in case there was a misunderstanding i wasn't criticising your building of the app, i know you've done the best you can =)

**Comment by stujones11** [comment]
@paramat api 14 is currently the minimum supported version. The actual 'target' api something different again, it just requires the we request permissions for access to the internal storage so MoNTE48's java changes would need to be backported for a playstore update of 0.4.17.x-?

My hunch is that the stl issues only effect api < 16 and if so, we'd need to up the minimum target to 16.

**Comment by pgimeno** [comment]
Why is mini-gmp disabled for Android? Does enabling it solve this issue?

https://github.com/minetest/minetest/blob/d994f7ca5f75431da8e17fd6762209c404d89482/src/util/srp.cpp#L49-L53

**Comment by nerzhul** [comment]
Full GMP is used instead of mini-gmp. If i remember this was just to have a proper packaging.

**Comment by paramat** [comment]
So what's the status of this? What needs to be done?

**Comment by nerzhul** [comment]
@paramat i published the stable-0.4 APK to the play store in the beta program and have the confirmation it's android before 6. 10 users installed the new apk using the beta program and the two crashes are on android 5 & 4.3

![capture d ecran de 2019-01-04 14-11-47](https://user-images.githubusercontent.com/119752/50689675-f402b600-102a-11e9-939a-94aaff27b2be.png)

**Comment by nerzhul** [comment]
@pgimeno i will try to generate a stable-0.4 apk with mini-gmp instead of gmp to see if there are issues

**Comment by nerzhul** [comment]
Fixed both in stable-0.4 and master
